alf/environments/suite_carla.py

```diff
@@ -1037,38 +1114,94 @@ def get_current_time_step(self, current_frame):
         obs['speed'] = np.float32(
             np.sqrt(np.float32(v.x * v.x + v.y * v.y + v.z * v.z)))
         self._current_distance = np.linalg.norm(obs['goal'])
+
+        prev_loc = _to_numpy_loc(self._prev_location)
+        curr_loc = _to_numpy_loc(self._actor.get_location())
+
         reward = 0.
         discount = 1.0
-        info = {}
-        info['success'] = np.float32(0.0)
+        info = OrderedDict(success=np.float32(0.0), collision=np.float32(0.0))
+
+        # When the previous episode ends because of stucking at a collision with
+        # another vehicle, it may get an additional collision event in the new frame
+        # because the relocation of the car may happen after the simulation of the
+        # moving. So we ignore the collision at the first step.
+        self._collision = not np.all(
+            obs['collision'] == 0) and not self._is_first_step
+        if self._collision and not self._prev_collision:
+            # We only report the first collision event among contiguous collision
+            # events.
+            info['collision'] = np.float32(1.0)
+            logging.info("actor=%d frame=%d COLLISION" % (self._actor.id,
+                                                          current_frame))
+            self._collision_loc = curr_loc
+            self._collision_frame = current_frame
+            # We don't want the collision penalty to be too large if the player
+            # cannot even get enough positive moving reward. So we cap the penalty
+            # at ``max(0., self._episode_reward)``
+            reward += max(self._collision_reward,
+                          -max(0., self._episode_reward))
 
         if self._fail_frame is None:
             step_type = ds.StepType.FIRST
             max_frames = math.ceil(
-                self._current_distance / self._min_speed / self._delta_seconds)
+                self._route_length / self._min_speed / self._delta_seconds)
             self._fail_frame = current_frame + max_frames
```

Comment:
Given the change, fail_frame should just be max_frames?

Reply:
updated